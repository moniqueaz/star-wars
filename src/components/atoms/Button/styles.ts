import styled, { css, DefaultTheme } from 'styled-components';

type ButtonProps = {
  color: 'light' | 'dark',
  size: 'small' | 'normal' | 'full',
}

const wrapperModifiers = {
  small: () => css`
    width: auto;
  `,
  normal: (theme: DefaultTheme) => css`
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  full: () => css`
    width: 100%;
  `,
  light: (theme: DefaultTheme) => css`
    background-color: ${theme.colors.white};
    color: ${theme.colors.black};
  `,
  dark: (theme: DefaultTheme) => css`
    background-color: ${theme.colors.primary};
    color: ${theme.colors.white};
  `,
};

export const Button = styled.button<ButtonProps>`
  ${( { theme, size, color } ) => css`
    border: none;
    box-shadow: none;
    border: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};
    font-size: ${theme.font.sizes.medium};
    ${wrapperModifiers[color](theme)};
    ${wrapperModifiers[size](theme)};
  `}
`;
