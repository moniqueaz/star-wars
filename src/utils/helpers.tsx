export const fetcher = (url: string) => fetch(url).then(res => res.json());

export const getRandomInteger = (min: number, max: number) => {
  return Math.floor(Math.random() * (max - min)) + min;
};
