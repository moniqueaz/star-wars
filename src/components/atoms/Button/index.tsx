import * as S from './styles';

type ButtonProps = {
  children: React.ReactNode,
  size?: 'small' | 'normal' | 'full',
  color?: 'light' | 'dark',
  callback?: () => any
}
const Button = ( { children, size = 'normal', color = 'light', callback }: ButtonProps) => (
  <S.Button size={size} color={color} onClick={callback}>
    {children}
  </S.Button>
);

export default Button;
