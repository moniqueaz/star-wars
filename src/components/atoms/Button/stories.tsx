import { Story, Meta } from '@storybook/react/types-6-0';
import Button from '.';

export default {
  title: 'Atoms/Button',
  component: Button,
} as Meta;

export const Default: Story = args => {
  return <Button {...args}>Button Text</Button>;
};

export const Small: Story = () => {
  return <Button size="small">Button Text</Button>;
};

export const Full: Story = () => {
  return <Button size="full">Button Text</Button>;
};

export const Dark: Story = () => {
  return <Button color="dark">Button Text</Button>;
};
