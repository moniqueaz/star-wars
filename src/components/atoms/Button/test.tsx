import { renderWithTheme } from 'utils/tests/helpers';

import Button from '.';

describe('<Button />', () => {
  it('should render', () => {
    const { container } = renderWithTheme(<Button>Button</Button>);
    expect(container.querySelector('button')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  } );

  it('should render small', () => {
    const { container } = renderWithTheme(<Button size="small">Button</Button>);
    expect(container.querySelector('button')).toHaveStyle( { width: 'auto' } );
  } );

  it('should render normal', () => {
    const { container } = renderWithTheme(<Button size="normal">Button</Button>);
    expect(container.querySelector('button')).toHaveStyle( { padding: '0.8rem 3.2rem' } );
  } );

  it('should render large', () => {
    const { container } = renderWithTheme(<Button size="full">Button</Button>);
    expect(container.querySelector('button')).toHaveStyle( { width: '100%' } );
  } );

  it('should render dark', () => {
    const { container } = renderWithTheme(<Button color="dark">Button</Button>);
    expect(container.querySelector('button')).toHaveStyle( { ['background-color']: '#1A4173' } );
  } );

  it('should render light', () => {
    const { container } = renderWithTheme(<Button color="light">Button</Button>);
    expect(container.querySelector('button')).toHaveStyle( { ['background-color']: '#f1f1f1' } );
  } );
} );
