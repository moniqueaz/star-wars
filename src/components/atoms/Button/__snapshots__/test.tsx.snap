// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render 1`] = `
.c0 {
  border: none;
  box-shadow: none;
  border: 0.4rem;
  padding: 0.8rem;
  font-size: 1.6rem;
  background-color: #f1f1f1;
  color: #333;
  padding: 0.8rem 3.2rem;
}

<button
  class="c0"
  color="light"
>
  Button
</button>
`;
